File: growsafe_investments/lib/services/api_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;

class ApiService {
  static const String baseUrl = 'http://127.0.0.1:8000/api/auth';

  static Future<Map<String, dynamic>> signup(String username, String email, String password, String confirmPassword) async {
    final url = Uri.parse('$baseUrl/signup/');
    final response = await http.post(
      url,
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({
        'username': username,
        'email': email,
        'password': password,
        'confirm_password': confirmPassword,
      }),
    );
    return jsonDecode(response.body);
  }

  static Future<Map<String, dynamic>> login(String username, String password) async {
    final url = Uri.parse('$baseUrl/login/');
    final response = await http.post(
      url,
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({
        'username': username,
        'password': password,
      }),
    );
    return jsonDecode(response.body);
  }

  static Future<Map<String, dynamic>> logout(String token) async {
    final url = Uri.parse('$baseUrl/logout/');
    final response = await http.post(
      url,
      headers: {
        'Content-Type': 'application/json',
        'Authorization': 'Bearer $token',  // Fixed to Bearer
      },
    );
    return jsonDecode(response.body);
  }

  static Future<Map<String, dynamic>> getProfile(String token) async {
    final url = Uri.parse('$baseUrl/profile/');
    final response = await http.get(
      url,
      headers: {
        'Content-Type': 'application/json',
        'Authorization': 'Bearer $token',
      },
    );
    return jsonDecode(response.body);
  }

  static Future<Map<String, dynamic>> refreshToken(String refreshToken) async {
    final url = Uri.parse('$baseUrl/token/refresh/');  // Fixed URL
    final response = await http.post(
      url,
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({'refresh': refreshToken}),
    );
    return jsonDecode(response.body);
  }

  static Future<Map<String, dynamic>> deposit(String token, double amount) async {
    final url = Uri.parse('$baseUrl/deposit/');
    final response = await http.post(
      url,
      headers: {
        'Content-Type': 'application/json',
        'Authorization': 'Bearer $token',
      },
      body: jsonEncode({'amount': amount}),
    );
    return jsonDecode(response.body);
  }

  static Future<Map<String, dynamic>> withdraw(String token, double amount) async {
    final url = Uri.parse('$baseUrl/withdraw/');
    final response = await http.post(
      url,
      headers: {
        'Content-Type': 'application/json',
        'Authorization': 'Bearer $token',
      },
      body: jsonEncode({'amount': amount}),
    );
    return jsonDecode(response.body);
  }

  static Future<Map<String, dynamic>> invest(String token, String name, double amount, double dailyReturnRate) async {
    final url = Uri.parse('$baseUrl/invest/');
    final response = await http.post(
      url,
      headers: {
        'Content-Type': 'application/json',
        'Authorization': 'Bearer $token',
      },
      body: jsonEncode({
        'name': name,
        'amount': amount,
        'daily_return_rate': dailyReturnRate,
      }),
    );
    return jsonDecode(response.body);
  }
}

File: reactfrontend/src/services/apiService.ts
import axios from 'axios';
import { AuthResponse, User } from '../models/types';

const BASE_URL = 'http://127.0.0.1:8000/api/auth';

const api = axios.create({
  baseURL: BASE_URL,
  headers: { 'Content-Type': 'application/json' },
});

export const signup = async (username: string, email: string, password: string, confirmPassword: string) => {
  const response = await api.post('/signup/', { username, email, password, confirm_password: confirmPassword });
  return response.data as AuthResponse;
};

export const login = async (username: string, password: string) => {
  const response = await api.post('/login/', { username, password });
  return response.data as AuthResponse;
};

export const logout = async (token: string) => {
  const response = await api.post('/logout/', {}, { headers: { Authorization: `Bearer ${token}` } });
  return response.data;
};

export const getProfile = async (token: string) => {
  const response = await api.get('/profile/', { headers: { Authorization: `Bearer ${token}` } });
  const rawData = response.data;
  const parsedData: User = {
    ...rawData,
    total: Number(rawData.total), // Convert to number
    totalDeposit: Number(rawData.totalDeposit),
    totalWithdraw: Number(rawData.totalWithdraw),
    dailyEarnings: Number(rawData.dailyEarnings),
    investments: rawData.investments.map((inv: any) => ({
      ...inv,
      amount: Number(inv.amount),
      dailyReturnRate: Number(inv.dailyReturnRate),
    })),
  };
  return parsedData;
};

export const deposit = async (token: string, amount: number) => {
  const response = await api.post('/deposit/', { amount }, { headers: { Authorization: `Bearer ${token}` } });
  return response.data;
};

export const withdraw = async (token: string, amount: number) => {
  const response = await api.post('/withdraw/', { amount }, { headers: { Authorization: `Bearer ${token}` } });
  return response.data;
};

export const invest = async (token: string, name: string, amount: number, dailyReturnRate: number) => {
  const response = await api.post('/invest/', { name, amount, daily_return_rate: dailyReturnRate }, { headers: { Authorization: `Bearer ${token}` } });
  return response.data;
};

export const refreshToken = async (refresh: string) => {
  const response = await api.post('/token/refresh/', { refresh });
  return response.data as { access: string };
};